### ISSUE TITLE
Cannot override a service exported with typescript file

### ISSUE THREAD

**Issue by Gaurav0** [typed-ember/ember-cli-typescript#562]
### Please paste the output of `ember -v` here
ember-cli: 3.7.1
node: 8.12.0
os: linux x64

### Please paste the output of `tsc -v` here
tsc -v
The program 'tsc' is currently not installed. You can install it by typing:
sudo apt install node-typescript

### Please paste your `tconfig.json` and `tslint.json` or `eslint.json` (if applicable) below
<details><summary><b>My tsconfig.json</b></summary><pre>

```
{
  "compilerOptions": {
    "target": "es2017",
    "allowJs": true,
    "moduleResolution": "node",
    "allowSyntheticDefaultImports": true,
    "noImplicitAny": true,
    "noImplicitThis": true,
    "alwaysStrict": true,
    "strictNullChecks": true,
    "strictPropertyInitialization": true,
    "noFallthroughCasesInSwitch": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noImplicitReturns": true,
    "noEmitOnError": false,
    "noEmit": true,
    "inlineSourceMap": true,
    "inlineSources": true,
    "baseUrl": ".",
    "module": "es6",
    "paths": {
      "repro-ajax-ts/tests/*": [
        "tests/*"
      ],
      "repro-ajax-ts/*": [
        "app/*"
      ],
      "*": [
        "types/*"
      ]
    }
  },
  "include": [
    "app/**/*",
    "tests/**/*",
    "types/**/*"
  ]
}
```
  
</pre></details>

### What are instructions we can follow to reproduce the issue?

1. ember new repro-ajax-ts
2. ember install ember-cli-typescript@next
3. Bump ember-ajax to 4.0.2
4. Create a service named ajax.js that extends ajax service from ember-ajax

##### Reproduction Case

Reproduction: https://github.com/Gaurav0/repro-ajax-ts (Run ember test)

### Now about that bug. What did you expect to see?

Expected test to pass; i.e. was able to override ajax service from addon

### What happened instead?

Test fails; ajax service is the one from the addon; not the app.

TLDR: if an addon exports a service using a .ts file in its app directory, an app using it cannot override it using a .js file

Related issue: https://github.com/ember-cli/ember-ajax/issues/428

**Comment by chriskrycho** [comment]
This is documented in the RC release notes. I’ll make sure we document it clearly in the docs proper as we roll toward release.

**Comment by alexlafroscia** [comment]
Hey @chriskrycho -- where is that documented in the release notes? I'm looking through the changelogs attached to each Github release and don't see this referenced anywhere... I'm not sure how addon authors are supposed to handle this.

**Comment by chriskrycho** [comment]
Sorry, it was in [the release notes for the first v2 beta](
https://github.com/typed-ember/ember-cli-typescript/releases/tag/v2.0.0-beta.1).

Addon authors should use .js, not .ts, in `app` (and for the sake of good imports, we generally recommend doing nothing except re-exporting there anyway).

**Comment by alexlafroscia** [comment]
Ahh, I see! Thanks for the clarification there, I think I can fix this in `ember-ajax` then.

**Comment by chriskrycho** [comment]
Interesting – good suggestion. @dfreeman will you have time to look at that next week? If not, I'll try to carve out some time to that effect.

**Comment by dfreeman** [comment]
We talked about doing that a while back, but the issue is that there's no hook to get the addon's app tree (`treeForApp` gives us ember-cli-typescript's _own_ tree), which is the same reason we can't transpile it in the first place.

We may be able to do some sort of best-effort check just by crawling `app` in development, though—I'll think about that a bit.

**Comment by chriskrycho** [comment]
@dfreeman any further thoughts on whether we should tackle this via attempting to be smart in our development process? The alternative is simply to document it clearly in the guides we're about to put together for this.

**Comment by Gaurav0** [comment]
It could be possible to modify ember-cli's build process to handle this correctly or to provide an appropriate hook? You may need to write an RFC.

**Comment by chriskrycho** [comment]
Possibly, but I honestly don't see a lot of value in doing that: in my opinion it's an anti-pattern to do much of anything besides re-exports in `app` space (and *something*—either module unification, template imports, or something solving that problem) should serve to *eliminate* that over time anyway, since a more standard approach to resolution will be in place.

As such the only thing I'm possibly in favor of (not speaking for the rest of the team) is a temporary workaround to *warn* users.

**Comment by dfreeman** [comment]
@chriskrycho I spiked out a simple warning the other day that just crawls `app` if it exists for addons with `isDevelopingAddon()` set. Will open a PR with that change in the next day or two, which should solve the primary issue here.

> in my opinion it's an anti-pattern to do much of anything besides re-exports in `app` space

:100:% agreed.
